src/MSMToolsCluster.jl

```diff
@@ -0,0 +1,259 @@
+            for j in transition_states
+                chi[j, :] .+= h[j] .* chi[i, :]
+            end
+        end
+    end
+
+    nmeta = count(x -> x != 0, sum(chi, dims=1))
+    if nmeta < m
+        throw(error("Requested $m metastable states, but transition matrix only has $nmeta. " *
+                "Try requesting fewer metastable states."))
+    end
+
+    return chi
+end
+
+# Reimplementation of PCCA+ spectral clustering method with optimized memberships (assumes fully connected transition matrix)
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools/analysis/dense/_pcca.py#L197
+function pcca_connected(P, n, pi=nothing)
+    labels = connected_sets(P)
+    n_components = length(labels)
+    if n_components > 1
+        throw(error("Unable to use pcca_connected, transition matrix is not fully connected."))
+    end
+
+    if pi === nothing
+        pi = stationary_distribution(P)
+    else
+        if size(pi, 1) != size(P, 1)
+            throw(error("Stationary distribution must span the entire state space but got $(size(pi, 1)) states " *
+                                "instead of $(size(P, 1))."))
+        end
+        pi ./= sum(pi)  # normalization
+    end
+
+    eigenvalues, eigenvectors = eigen(P)
+    indices = sortperm(eigenvalues, rev=true)
+    evecs = eigenvectors[:, indices]
+
+    # orthonormalize
+    for i in 1:n
+        evecs[:, i] /= sqrt(dot(evecs[:, i] .* pi, evecs[:, i]))
+    end
+
+    # make 1st eigenvector positive
+    evecs[:, 1] = abs.(evecs[:, 1])
+
+    if !all(isreal.(evecs))
+        println("Transition matrix has complex eigenvectors. Forcing eigenvectors to be real (see DeepTime note).")
+    end
+    evecs = real.(evecs)
+
+    # start with initial PCCA+ solution (can contain negative memberships)
+    chi, rot_matrix = pcca_connected_isa(evecs, n)
+
+    # optimizes PCCA+ rotation matrix s.t. memberships are always nonnegative
+    rot_matrix = opt_soft(evecs, rot_matrix, n)
+    memberships = evecs[:, :] * rot_matrix
+
+    # force memberships to be in range [0,1] (due to numerical errors)
+    memberships = max.(memberships, 0.0)
+    memberships = min.(memberships, 1.0)
+
+    for i in 1:size(memberships, 1)
+        memberships[i, :] ./= sum(memberships[i, :])
+    end
+
+    return memberships
+end
+
+# Reimplementation of PCCA+ spectral clustering method using the inner simplex algorithm.
+# https://github.com/deeptime-ml/deeptime/blob/a6ac0b93a55d688fe8f3af119680105763366220/deeptime/markov/tools/analysis/dense/_pcca.py#L14
+function pcca_connected_isa(eigenvectors, num_clusters)
+    n, m = size(eigenvectors)
+    if num_clusters > m
+        throw(ArgumentError("Cannot cluster the ($n x $m) eigenvector matrix to $num_clusters clusters."))
+    end
+
+    # Check if (ONLY) 1st eigenvector is constant
+    diffs = abs.(maximum(eigenvectors, dims=1) .- minimum(eigenvectors, dims=1))
+    if diffs[1] >= 1e-6
```

Comment:
It might be a good idea to set the tolerance as a keyword argument of the function so it can be switched easily for testing

Reply:
fixed in d1944f227a6e589496531e133bbd8feaf8c86dca